consensus/safety-rules/src/tests/suite.rs

```diff
@@ -600,3 +607,144 @@ fn test_sign_proposal_with_early_preferred_round(func: Callback) {
         Error::InvalidQuorumCertificate("Preferred round too early".into())
     );
 }
+
+fn test_uninitialized_signer(func: Callback) {
+    // Testing for an uninitialized Option<ValidatorSigner>
+
+    let (mut safety_rules, signer) = func();
+
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+    let err = safety_rules.update(a1.block().quorum_cert()).unwrap_err();
+    assert_eq!(err, Error::NotInitialized);
+
+    safety_rules.initialize(&proof).unwrap();
+    assert!(safety_rules.update(a1.block().quorum_cert()).is_ok());
+}
+
+#[allow(dead_code)]
+fn test_missing_validator_key(func: Callback) {
+    // Tests for fetching a missing validator key from persistent storage.
+    // Ideally safety rules should panic
+
+    let (mut safety_rules, signer) = func();
+    let (proof, genesis_qc) = make_genesis(&signer);
+    let round = genesis_qc.certified_block().round();
+
+    safety_rules.initialize(&proof).unwrap();
+
+    let a1 = test_utils::make_proposal_with_qc(round + 1, genesis_qc, &signer);
+
+    let mut next_epoch_state = EpochState::empty();
+    next_epoch_state.epoch = 1;
+    let rand_signer = ValidatorSigner::random([0xfu8; 32]);
+    next_epoch_state.verifier =
+        ValidatorVerifier::new_single(signer.author(), rand_signer.public_key());
+    let a2 = test_utils::make_proposal_with_parent_and_overrides(
+        vec![],
+        round + 2,
+        &a1,
+        Some(&a1),
+        &signer,
+        Some(1),
+        Some(next_epoch_state),
+    );
+    assert!(safety_rules.update(a2.block().quorum_cert()).is_err());
```

Comment:
unwrap_err please

Reply:
also its not really clear to me why we don't just make a1 with the rand signer and eliminate a2? do we need the epoch change?

Reply:
This is because the whole tests are about key reconciliation during epoch change, requiring the test code to actually trigger "update(...)". 
Making a a legal a1 is necessary to pass the QC verification in the 1st line of update(...), and thus enabling the test code to reach the part I'm trying to test here -- the validator tries to grab a its new consensus key but fails due to its acct is not in the set.

Reply:
Per the comments given in the PR, we should be going through reconciliation every time we start a clean safety rules and hit initialize.